web/vibe/web/rest.d

```diff
@@ -1444,12 +1445,16 @@ private HTTPServerRequestDelegate jsonMethodHandler(alias Func, size_t ridx, T)(
 					} else static if (sparam.kind == ParameterKind.wholeBody) {
 						try v = deserializeJson!PT(req.json);
 						catch (JSONException e) enforceBadRequest(false, e.msg);
+						catch (SerializationLogicError e) enforceBadRequest(false, e.msg);
 					} else static if (sparam.kind == ParameterKind.body_) {
 						try {
 							if (auto pv = fieldname in req.json)
 								v = deserializeJson!PT(*pv);
-						} catch (JSONException e)
+						} catch (JSONException e) {
 							enforceBadRequest(false, e.msg);
+						} catch (SerializationLogicError e) {
```

Comment:
What's the point of introducing a new exception type if the handling is exactly the same ?
Also, it's an `Exception`, not an `Error`. In D, `XXXError` are classes derived from `Error` by convention.

Reply:
> What's the point of introducing a new exception type if the handling is exactly the same ?

Well a custom exception class is here because it's being thrown from a generic deserialization routine (not a JSON one). The same logic works well on a `BsonSerializer` or any other custom serializer.
So the exception definitely should not be a `JSONException`.

Reply:
> Also, it's an Exception, not an Error. In D, XXXError are classes derived from Error by convention.

Ok then. I'll rename it into `SerializationLogicException`. Thanks :)

Reply:
> So the exception definitely should not be a `JSONException`.

Except, any client will not want to catch both `JSONException` and `SerializationLogicException`.  So we're forcing callers to handle 2 exception type for an internal detail. That's a pretty bad API IMO.

Reply:
> > So the exception definitely should not be a `JSONException`.
> 
> Except, any client will not want to catch both `JSONException` and `SerializationLogicException`. So we're forcing callers to handle 2 exception type for an internal detail. That's a pretty bad API IMO.

Agree. Do you think it'll be a good decision to wrap the underlying exception everywhere earlier?
```d
...
try {
  return ser.deserializeValue!(T)(...);
} catch (Exception e) {
  throw new SerializationLogicException(...);
}
```

Or maybe we should introduce the same enum-handling logic in every serializer implementation?

Reply:
I'd say the logic should go in the serializer.
It's not a great solution, but it's the best option I can think of.

Bear in mind that someone could very well be calling `std.conv.to` and get a `ConvException` so you can never really rely on the exception type anyway.

By the way catch-and-rethrow is quite inefficient, not something we'd want in a high performance web server.

Reply:
Ok then let's place it into serializer and throw well-known `JSONException` manually. I agree it's an acceptable solution which will allow us not to check the whole codebase against my changes.

I'll return with my changes soon. Thanks )

Reply:
Thanks for taking care of this. Let me know when this is ready for another review.

Reply:
@Geod24 Hey, I copied the logic into a serializer. Now the `JSONSerializer` is responsible for parsing enum values while any other implementation will use the old generic logic.
The bad news is that there is also a `@byName` handling logic which was also copied =(

Have you got an idea how to make the solution better without bloating the serializer "interface"?